[CBRD-26214] commit with an open cursor

hyunikn opened the pull request:
<http://jira.cubrid.org/browse/CBRD-26214>

### Purpose

PL/CSQL 커서가 열린 상태에서 COMMIT 을 할 때 커서 FETCH에서 cub_server 가 crash 하는 문제 해결. 
COMMIT 을 통해 현재 트랜잭션에 속한 커서의 자원이 해제되는 과정이 문제 유발.

### Implementation

서버사이드 JDBC에서 query prepare 를 할 때 holdable 옵션을 적용함으로서 COMMIT 문이 커서 자원 해제를 하지 않도록 수정

### Remarks

SP 호출에 해당하는 call stack 이 pop 될 때 해당 call stack에 속한 커서가 모두 close 되므로 
holdable cursor로 바뀌어도 stack 이 pop 될 때 관련 자원은 해제된다.

hyunikn committed:
code style

hyunikn commented:
/run all

hyunikn commented:
Shell 테스트에서 실패한 20건에 대한 실패 원인은 다음과 같습니다. 

    // unstable: fail on CI and success on private dev VM
    shell/_06_issues/_11_1h/bug_bts_4633/cases/bug_bts_4633.sh

    // unstable: Excel
    shell/_06_issues/_11_1h/bug_bts_5492/cases/bug_bts_5492.sh
    shell/_06_issues/_11_1h/bug_bts_5492/cases/bug_bts_5492.sh
    shell/_06_issues/_11_1h/bug_bts_5492/cases/bug_bts_5492.sh
    shell/_06_issues/_11_1h/bug_bts_5492/cases/bug_bts_5492.sh

    // unstable: Excel
    shell/_06_issues/_12_2h/bug_bts_8379/cases/bug_bts_8379.sh
    shell/_06_issues/_12_2h/bug_bts_8379/cases/bug_bts_8379.sh
    shell/_06_issues/_12_2h/bug_bts_8379/cases/bug_bts_8379.sh
    shell/_06_issues/_12_2h/bug_bts_8379/cases/bug_bts_8379.sh

    // unstable: Excel
    shell/_06_issues/_12_2h/bug_bts_9496/cases/bug_bts_9496.sh
    shell/_06_issues/_12_2h/bug_bts_9496/cases/bug_bts_9496.sh
    shell/_06_issues/_12_2h/bug_bts_9496/cases/bug_bts_9496.sh
    shell/_06_issues/_12_2h/bug_bts_9496/cases/bug_bts_9496.sh

    // unstable: Excel
    shell/_06_issues/_25_1h/cbrd_25992/cases/cbrd_25992.sh
    shell/_06_issues/_25_1h/cbrd_25992/cases/cbrd_25992.sh
    shell/_06_issues/_25_1h/cbrd_25992/cases/cbrd_25992.sh
    shell/_06_issues/_25_1h/cbrd_25992/cases/cbrd_25992.sh

    // unstable: Excel
    shell/_06_issues/_25_1h/cbrd_26020/cases/cbrd_26020.sh

    // another issue: Excel
    shell/_08_shard/_17_shard_cas_log/cases/_17_shard_cas_log.sh

    // 조사 중
    shell/_29_features_920/issue_10987_volume/_03_volume_config/cases/_03_volume_config.sh

ctshim commented on src/sp/pl_query_cursor.cpp:
```diff
@@ -87,6 +87,12 @@ namespace cubpl
+	if (m_query_entry->list_id == NULL)
+	  {
+	    int tran_index = LOG_FIND_THREAD_TRAN_INDEX (m_thread);
+	    m_query_entry = qmgr_get_query_entry (m_thread, m_query_id, tran_index);
```
이곳을 포함해서 3군데에서 동일한 코드가 추가 되었습니다.
이 추가된 코드 전체에 대해서 
`(m_query_entry->list_id == NULL) `인 경우에 대한 에러 처리가 필요할 것 같습니다.
현재는 이런 경우에 core dump가 발생하게 될 것입니다.

qmgr_get_query_entry() 함수가 절대로 널을 리턴하지 않을 것이라는 걸 보장 받을 수 있는 건가요?

hyunikn replied:
코드 분석과 테스트를 통해서 그 자리에서 NULL 체크와 에러 처리를 해야 한다는 것을 알게 되었습니다. 

서버사이드 JDBC에서 query prepare  가 모두 holdable 로 바뀐 상황에서 COMMIT 이후에 qmgr_get_query_entry () 호출은 세션 영역에 저장되어 있던 (holdable) 정보를 load 해서 새로운 query entry 를 만들어 transaction entry 에 추가하고 그것을 돌려 줍니다. 다시 말해 세션 영역으로부터의 정보 load 가 실패하지 않으면 (qmgr_get_query_entry() 에서 xsession_load_query_entry_info() 호출이 실패하지 않는 한) NULL이 되지 않을 것입니다. 어떤 이유로 정보 load 가 실패할 가능성도 배제할 수 없습니다. 그 경우에는 NULL 이 될 것입니다. 
그런데, COMMIT 이 아니라 ROLLBACK의 경우에는 무조건 NULL 돌려 주게 됩니다. CBRD-26214 예제에서 COMMIT 대신 ROLLBACK 으로 바꾼 아래 예제에서는 qmgr_get_query_entry() 가 NULL 을 리턴합니다. 
```
create or replace procedure poo as
    cursor c is select /*+ no_parallel_heap_scan */ i from ttt;
    v int;
begin
    open c;
    for k in 1 .. 10000 loop
        fetch c into v;
        if (k mod 2000 = 0) then
            rollback;
        end if;
    end loop;
    close c;
end;
```
이는 ROLLBACL 실행 중에 세션 영역 중에 있는 정보도 지워서 (qmgr_clear_trans_wakeup() 에서 xsession_clear_query_entry_info() 호출) query entry 를 새로 만들 정보가 남아 있지 않기 때문인 것 같습니다. 

NULL 체크와 에러 처리 과정을 추가하겠습니다.

hyunikn replied:
null 체크 과정을 추가했습니다.

mhoh3963 commented on src/sp/pl_query_cursor.cpp:
```diff
@@ -135,7 +141,12 @@ namespace cubpl
+	if (m_query_entry->list_id == NULL)
+	  {
+	    int tran_index = LOG_FIND_THREAD_TRAN_INDEX (m_thread);
+	    m_query_entry = qmgr_get_query_entry (m_thread, m_query_id, tran_index);
```
m_query_entry->list_id 가 NULL인 경우에 m_query_entry를 free 가 필요해보입니다.

hyunikn replied:
여기서 free 를 한다면 double free 하게 될 가능성이 높아 보입니다. 

COMMIT 이나 ROLLBACK 을 실행하는 과정에서 qmgr_clear_trans_wakeup() 가 qmgr_free_query_entry() 를 호출하게 되는데, 이는 실제로 query entry 에 대해서 free 를 호출하는 것이 아니고, transaction entry 의 free_query_entry_list_p 로 옮겨 놓습니다. 이 free list 에 저장된 entry 들에 대해 나중 어느 시점에 한꺼번에 free 를 호출하는 것 같습니다. 이 경우 144 라인에서의 m_query_entry 참조는 그 free list 로 옮겨져 있는 query entry 에 대한 참조라고 보여집니다. 
단, 144 라인에서의 m_query_entry 참조에 앞서 free list 전체에 대해 실제로 메모리 해제가 일어났을 가능성은 없는지, 즉, 해제된 메모리 영역을 참조할 수도 있는지 좀 더 확인해 봐야 할 것 같습니다.

mhoh3963 replied:
qmgr_get_query_entry() 대신 session_load_query_entry_info() 사용하는 것은 어떨까요 ?

hyunikn replied:
> qmgr_get_query_entry() 대신 session_load_query_entry_info() 사용하는 것은 어떨까요 ?

코드를 살펴보니 qmgr_get_query_entry() 에서 내부적으로 session_load_query_entry_info() 를 
호출하고 있습니다. 
그 외에도 transaction entry 의 query entry list 에 새로 해당 entry 를 끼워넣는 일도 하고 있는데, 이 일도 필요해 보입니다. 이 것을 하지 않으면 4000 번째에서 다시 commit 을 실행할 때 그 때까지의 cursor에 변경 내용을 세션 영역으로 다시 복사해 둘 수 없겠습니다. (참고: qmgr_clear_trans_wakeup() 에서 xsession_store_query_entry_info() 호출) 

qmgr_get_query_entry() 사용에는 문제가 없어 보입니다.

hyunikn replied:
> 단, 144 라인에서의 m_query_entry 참조에 앞서 free list 전체에 대해 실제로 메모리 해제가 일어났을 가능성은 없는지, 즉, 해제된 메모리 영역을 참조할 수도 있는지 좀 더 확인해 봐야 할 것 같습니다.

추가로 분석을 해보니, query entry 는 한 번 만들어지면 DBMS가 내려가지 않는 한 메모리에서 free 되는 일은 없는 것 같습니다. 

transaction entry 의 free_query_entry_list_p 의 entry 들을 메모리 해제하는 코드는 qmgr_free_tran_entries() 밖에 없는 것 같고, 이 함수는 qmgr_finalize() 에서만 호출됩니다. 

transaction entry의 (active) query entry list 와 free query entry list 를 왕복해 가며 재사용 되는 것 같습니다. 
(참고: qmgr_allocate_query_entry() 와 qmgr_free_query_entry())

다만, 144 라인에서 참조되는 m_query_entry 가 이 cubpl::query_cursor 가 만들어질 당시 query 의 것이 아니라 commit이나 rollback 에 의해 free list 에 한번 갔다가 다른 query에 할당된 query entry 일 가능성도 있어 보입니다. 이런 일이 발생할 수 있는지 검토해 보겠습니다.

hyunikn replied:
예상했던 문제가 다음 예제에서 재현됩니다. 

```
drop table if exists ttt;
create table ttt (i int);
insert into ttt select rownum from db_class a, db_class b, db_class c, db_class d limit 10000;

drop table if exists yyy;
create table yyy (i string);
insert into yyy select 'a'||rownum from db_class a, db_class b, db_class c, db_class d limit 10000;

create or replace procedure poo as
    cursor c is select /*+ recompile */ i from ttt;
    cursor d is select /*+ recompile */ i from yyy;
    v int;
    w string;
begin
    open c;
    fetch c into v;    -- default fetch size = 1000
    dbms_output.put_line('v=' || v);

    commit;
    open d;
    fetch d into w;

    for k in 2 .. 1001 loop
        begin
            fetch c into v;
            dbms_output.put_line('v=' || v);
        exception
        when others then
            dbms_output.put_line('exception for k=' || k || ': ' || sqlerrm);
        end;
    end loop;
end;

call poo();

결과
-------------------------
...
...
v=996
v=997
v=998
v=999
exception for k=1000: internal server error  //  java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
exception for k=1001: internal server error

Execute OK. (0.074001 sec) Committed. (0.000000 sec) 

1 command(s) successfully processed.

```
query cursor c open 할 때 할당되었던 query entry 가 commit 에 의해서 free list 에 갔다가 
open d 에 의해 query cursor d에 다시 할당됩니다. 
그런데, c 는 여전히 free list 에 갔다가 온 query entry 를 참조하기 때문에 발생하는 문제 입니다. 

query entry 를 할당할 때마다 일련번호를 하나씩 증가시키고
query cursor 는 할당받은 query entry 의 일련번호를 기억하고 있다가 
commit 등의 이유로 이것이 달라지면 다시 할당받는 방식으로 수정 하려고 합니다.

hyunikn replied:
위 내용 관련해서 수정을 추가했습니다.

hyunikn committed:
Merge branch 'develop' into cbrd-26214-commit-with-open-cursor-develop

hyunikn commented:
/run all

mhoh3963 commented:
@hyunikn 
log_commit_local()에서 호출하는 qmgr_clear_trans_wakeup()를 보면 아래와 같은 pl 관련 code가 있습니다.
즉, holdable이 아니여도, pl 수행 중에는 commit를 해도 clear되면 안될 것 같은데요. 확인부탁드립니다.

  while (query_p)
    {
      if (is_pl_session_running && query_p->query_status == QUERY_IN_PROGRESS)
        {
          query_p = query_p->next;
          continue;
        }

hyunikn committed:
. add alloc_no to struct qmgr_query_entry to identify its allocation during the recycling process
. let cubpl::query_cursors remember the alloc_no of their qmgr_query_entry because
  the entry can be reassigned to another query cursor
. add null checks to the result of qmgr_get_query_entry()

hyunikn committed:
Merge branch 'develop' into cbrd-26214-commit-with-open-cursor-develop

hyunikn commented:
/run all

hgryoo commented:
다음 두가지에 대해서 추가 확인 부탁드립니다.
- 다른 쿼리캐시 모듈에 의해 공유될 수 있는 경우: QUERY_CACHE 힌트가 사용되는 경우
- 결과 파일이 생성되지 않을 경우: `SELECT 1;` 또는 `SELECT * FROM db_class WHERE 0 <> 0;`

mhoh3963 commented:
@hyunikn 
사용자가 정상적으로 cursor를 close후에  closed된 cursor를 fetch하는 경우에 대해는 정상적으로 오류가 발생하는지 추가 확인부탁드립니다.

beyondykk9 commented on src/query/query_manager.c:
```diff
@@ -290,6 +290,11 @@ qmgr_allocate_query_entry (THREAD_ENTRY * thread_p, QMGR_TRAN_ENTRY * tran_entry
       tran_entry_p->free_query_entry_list_p = query_p->next;
       pthread_mutex_unlock (&tran_entry_p->mutex);
+      query_p->alloc_no++;
+      if (query_p->alloc_no == 0)
```
0을 체크한다는 것은 이전 alloc_no가 overflow라는 전제인데, alloc_no가 그 정도 값이 될 수 있나요?
overflow가 될 수 없다면, assert로 걸어놔야 할 것 같습니다.

hyunikn replied:
unsigned int 의 최대값이 40억이 넘어가니 그럴 확률은 없다고 봐도 무방할 것 같습니다. 
assert 로 바꿔두어도 좋을 것 같습니다.

hyunikn replied:
사실 alloc_no 가 0 이어도 문제는 없습니다. 
그래서 assert 도 불필요 합니다. 
그냥 삭제하겠습니다.

hyunikn committed:
Merge branch 'develop' into cbrd-26214-commit-with-open-cursor-develop

hyunikn commented:
> @hyunikn 사용자가 정상적으로 cursor를 close후에 closed된 cursor를 fetch하는 경우에 대해는 정상적으로 오류가 발생하는지 추가 확인부탁드립니다.

PL/CSQL 의 경우 끝까지 fetch 하거나 안 하거나에 상관없이 오류가 발생합니다. 
단, 이 에러 체크는 PL/CSQL SpLib 안에 커서에 해당하는 정의인 class Query 에서 체크하는 것 입니다. 
```
csql> create or replace procedure poo as
csql>     cursor c is select code from athlete;
csql>     v int;
csql> begin
csql>     open c;
csql>     loop
csql>         fetch c into v;
csql>         exit when c%notfound;
csql>     end loop;
csql>     close c;
csql>     fetch c into v;     -- fetch after closing
csql> end;
Execute OK. (0.632014 sec) Committed. (0.000000 sec) 

1 command(s) successfully processed.
csql> call poo();

In the command from line 1,

ERROR: Stored procedure execute error: 
  (line 11, column 5) tried to fetch values with an unopened cursor
```
```
csql> create or replace procedure poo as
csql>     cursor c is select code from athlete;
csql>     v int;
csql> begin
csql>     open c;
csql>     fetch c into v;
csql>     close c;
csql>     fetch c into v;     -- fetch after closing
csql> end;
Execute OK. (0.046001 sec) Committed. (0.000000 sec) 

1 command(s) successfully processed.
csql> call poo();

In the command from line 1,

ERROR: Stored procedure execute error: 
  (line 8, column 5) tried to fetch values with an unopened cursor
```

hyunikn commented:
> @hyunikn 사용자가 정상적으로 cursor를 close후에 closed된 cursor를 fetch하는 경우에 대해는 정상적으로 오류가 발생하는지 추가 확인부탁드립니다.

JSP의 경우 에러가 나지 않습니다. 아래 예제처럼 ResultSet.next() 를 끝까지 하거나 안 하거나에 상관없이 에러 나지 않습니다. 
서버 쪽의 cubpl::query_cursor 는 ResultSet.close() 에 닫히지 않고 SP 호출이 리턴하여 stack 이 pop 될 때 닫힙니다.
```
// create or replace procedure poo as language java name 'Test.test()';

import java.sql.*;

public class Test {
    public static void test() throws SQLException {
        final Connection conn = DriverManager.getConnection("jdbc:default:connection::");
        PreparedStatement pstmt = conn.prepareStatement("select code from athlete");
        ResultSet rset = pstmt.executeQuery();
        while (rset.next()) {
            ;
        }
        rset.close();

        rset.next();

        pstmt.close();
        conn.close();
    }
}

```
```
// create or replace procedure poo as language java name 'Test.test()';

import java.sql.*;

public class Test {
    public static void test() throws SQLException {
        final Connection conn = DriverManager.getConnection("jdbc:default:connection::");
        PreparedStatement pstmt = conn.prepareStatement("select code from athlete");
        ResultSet rset = pstmt.executeQuery();

        rset.next();
        rset.close();

        rset.next();

        pstmt.close();
        conn.close();
    }
}
```

JDBC 스펙에 따르면 close 된 ResultSet에 next 를 호출하면 SQLException 이 나야 하지만, 이 부분이 아직 구현되어 있지 않습니다.

hyunikn committed:
. remove a alloc_no zero check: it can be zero
. fix the definition of is_pl_session_running

hyunikn commented:
/run all

hyunikn committed:
minor fix

hyunikn commented:
/run all

hyunikn committed:
minor fix

hyunikn commented:
> 다음 두가지에 대해서 추가 확인 부탁드립니다.
> 
> * 다른 쿼리캐시 모듈에 의해 공유될 수 있는 경우: QUERY_CACHE 힌트가 사용되는 경우

아래 예제에 대해 테스트 해 보았는데, 특별한 이상은 없는 것 같습니다. 
call poo() 로 실행하고 ;info qcache 로 확인해 보면 query cache entry 가 잘 만들어져 있습니다. 
어떤 점을 주의해서 보면 좋을까요.

```
create or replace procedure poo as
    cursor c is select /*+ query_cache */ host_year from game;
    y int;
    k int := 0;
begin
    open c;
    loop
        fetch c into y;
        exit when c%notfound;
        if (k mod 2000 = 0) then
            commit;
        end if;
        k := k + 1;
    end loop;
    close c;
end;

```

hyunikn commented:
> 다음 두가지에 대해서 추가 확인 부탁드립니다.
> 
> * 결과 파일이 생성되지 않을 경우: `SELECT 1;` 또는 `SELECT * FROM db_class WHERE 0 <> 0;`

이 경우에 대해서도 에러 없이 실행됩니다. 
```
create or replace procedure poo as
    cursor c is select 1;
    y int;
begin
    open c;
    commit;
    fetch c into y;
    dbms_output.put_line('y=' || y);
    commit;
    close c;
end;

csql> call poo();

=== <Result of CALL Command in Line 1> ===

  Result              
======================
  NULL                

<DBMS_OUTPUT>
====
y=1
```
```
create or replace procedure poo as
    cursor c is SELECT * FROM db_class WHERE 0 <> 0;
    r db_class%rowtype;
begin
    open c;
    commit;
    fetch c into r;
    dbms_output.put_line('y=' || case when r.class_name is null then 'null' else r.class_name end);
    commit;
    close c;
end;

csql> call poo();

=== <Result of CALL Command in Line 1> ===

  Result
======================
  NULL

<DBMS_OUTPUT>
====
y=null
```

hgryoo commented:
> > 다음 두가지에 대해서 추가 확인 부탁드립니다.
> > 
> > * 다른 쿼리캐시 모듈에 의해 공유될 수 있는 경우: QUERY_CACHE 힌트가 사용되는 경우
> 
> 아래 예제에 대해 테스트 해 보았는데, 특별한 이상은 없는 것 같습니다. 
> call poo() 로 실행하고 ;info qcache 로 확인해 보면 query cache entry 가 잘 만들어져 있습니다. 
> 어떤 점을 주의해서 보면 좋을까요.
> 
> ```
> create or replace procedure poo as
>     cursor c is select /*+ query_cache */ host_year from game;
>     y int;
>     k int := 0;
> begin
>     open c;
>     loop
>         fetch c into y;
>         exit when c%notfound;
>         if (k mod 2000 = 0) then
>             commit;
>         end if;
>         k := k + 1;
>     end loop;
>     close c;
> end;
> 
> ```
> 
> 

다른 세션의 PL이나, 하나의 세션에서 동일한 쿼리 캐시를 사용하는 여러 커서를 생성했을 때, commit 시 자원 해제에 문제가 없는지 확인할 필요가 있을 것 같습니다.

hyunikn committed:
Merge branch 'develop' into cbrd-26214-commit-with-open-cursor-develop

hyunikn commented:
> 다른 세션의 PL이나, 하나의 세션에서 동일한 쿼리 캐시를 사용하는 여러 커서를 생성했을 때, commit 시 자원 해제에 문제가 없는지 확인할 필요가 있을 것 같습니다.

위 예제를 아래와 같이 query cache 를 사용할 여러 커서를 생성하는 식으로 바꿔서 세 건의 CSQL 세션에서 번갈아 실행해 보았습니다. 
valgrind 에서 memory leak 은 없었습니다. 
```
create or replace procedure poo as
    cursor c is select /*+ query_cache */ host_year from game;
    cursor d is select /*+ query_cache */ host_year from game;
    cursor e is select /*+ query_cache */ host_year from game;
    y int;
    k int := 0;
begin
    open c;
    open d;
    open e;
    loop
        fetch c into y;
        fetch d into y;
        fetch e into y;
        exit when c%notfound;
        if (k mod 2000 = 0) then
            commit;
        end if;
        k := k + 1;
    end loop;
    close c;
    close d;
    close e;
end;
```
```
==1467429== LEAK SUMMARY:
==1467429==    definitely lost: 0 bytes in 0 blocks
==1467429==    indirectly lost: 0 bytes in 0 blocks
==1467429==      possibly lost: 0 bytes in 0 blocks
==1467429==    still reachable: 40,022 bytes in 3 blocks
==1467429==         suppressed: 0 bytes in 0 blocks
```

hyunikn commented:
> * 결과 파일이 생성되지 않을 경우: `SELECT 1;` 또는 `SELECT * FROM db_class WHERE 0 <> 0;`

아래와 같이 TC 를 만들어 세 건의 CSQL에서 번갈아 실행하는 동안 cub_server에 valgrind를 걸어 테스트 해 봤습니다. 
메모리 릭은 발견되지 않았습니다. 

```
create or replace procedure poo as
    cursor c is SELECT * FROM db_class WHERE 0 <> 0;
    r db_class%rowtype;
begin
    open c;
    commit;
    fetch c into r;
    dbms_output.put_line('y=' || case when r.class_name is null then 'null' else r.class_name end);
    commit;
    close c;
end;

create or replace procedure koo as
    cursor c is select 1;
    y int;
begin
    open c;
    commit;
    fetch c into y;
    dbms_output.put_line('y=' || y);
    commit;
    close c;
end;
```

hyunikn commented:
> @hyunikn log_commit_local()에서 호출하는 qmgr_clear_trans_wakeup()를 보면 아래와 같은 pl 관련 code가 있습니다. 즉, holdable이 아니여도, pl 수행 중에는 commit를 해도 clear되면 안될 것 같은데요. 확인부탁드립니다.
> 
> while (query_p) { if (is_pl_session_running && query_p->query_status == QUERY_IN_PROGRESS) { query_p = query_p->next; continue; }

@mhoh3963 오프라인으로 답변 드렸으나 다른 리뷰어들도 참고하시라고 적습니다. 

본 이슈의 관심 대상인 SP 안에서 실행되는 query 에 대해 이 시점에 QUERY_IN_PROGRESS 되는 경우는 없습니다.
이 시점은 commit, rollback, error 로 인한 transaction 취소를 처리하는 도중이고 이 때 SP 가 실행하는 query 는 모두 QUERY_COMPLETED 상태 입니다. 

"select SP() from ..." 처럼 SP 를 호출하여 아직 실행 도중인 SELECT 문에 대해서 query entry 가 사라지지 않도록 하기 위한 코드입니다.

hyunikn committed:
Merge remote-tracking branch 'upstream/develop' into cbrd-26214-commit-with-open-cursor-develop

hyunikn commented:
/run all

hyunikn committed:
Merge branch 'develop' into cbrd-26214-commit-with-open-cursor-develop

hyunikn commented:
/run all

hyunikn commented:
실패한 shell test 는 다음과 같습니다. (중복제거)
```
shell/_04_misc/_04_index/itrack_10002/cases/itrack_10002.sh
shell/_04_misc/_04_index/itrack_10005/cases/itrack_10005.sh
shell/_06_issues/_11_1h/bug_bts_5492/cases/bug_bts_5492.sh
shell/_06_issues/_12_2h/bug_bts_8347/cases/bug_bts_8347.sh
shell/_06_issues/_12_2h/bug_bts_8379/cases/bug_bts_8379.sh
shell/_06_issues/_12_2h/bug_bts_9521_2/cases/bug_bts_9521_2.sh
shell/_06_issues/_24_1h/cbrd_25291/cases/cbrd_25291.sh
shell/_06_issues/_25_1h/cbrd_25992/cases/cbrd_25992.sh
shell/_08_shard/_17_shard_cas_log/cases/_17_shard_cas_log.sh
shell/_10_plcsql/cbrd_25619/cases/cbrd_25619.sh
shell/_27_aprium_qa/_01_i18n/issue_7737_prefixkey/_01_case_insensitive_02/cases/_01_case_insensitive_02.sh
shell/_35_cherry/issue_21654_server_side_loaddb/loaddb_CS/_35_cherry/issue_21506_online_index/cbrd_22477_unload/cases/cbrd_22477_unload.sh
shell/_38_fig/cbrd_24614/cases/cbrd_24614.sh
```
이 중에 develop branch (8dae125) 에서 이미 실패하고 있는 건(https://app.circleci.com/pipelines/github/CUBRID/cubrid/21170/workflows/1f58ea95-752f-4c32-9776-7db7469ed058/jobs/100598/tests)
을 제외하면 아래와 같습니다. 
```
shell/_04_misc/_04_index/itrack_10002/cases/itrack_10002.sh
shell/_06_issues/_12_2h/bug_bts_8347/cases/bug_bts_8347.sh
shell/_06_issues/_12_2h/bug_bts_9521_2/cases/bug_bts_9521_2.sh
shell/_10_plcsql/cbrd_25619/cases/cbrd_25619.sh
shell/_27_aprium_qa/_01_i18n/issue_7737_prefixkey/_01_case_insensitive_02/cases/_01_case_insensitive_02.sh
shell/_35_cherry/issue_21654_server_side_loaddb/loaddb_CS/_35_cherry/issue_21506_online_index/cbrd_22477_unload/cases/cbrd_22477_unload.sh
```
이 중 마지막 건을 제외하면 이미 unstable 로 리포트 된 것들이고, 
마지막  cbrd_22477_unload.sh 는 개인 개발 VM에서는 성공하므로 unstable 입니다. 

본 PR 변경에 의해 추가로 실패하는 shell TC 는 없어 보입니다.

hyunikn commented:
@jongmin-won 시간 관계 상 일단 머지 합니다. 나중에라도 문제가 보이시면 알려 주세요.
